### PR TITLE
[release] Upload & download blobs from Azure

### PR DESCRIPTION
**Pull request by khluu**
This supports the following functionality to enable Azure on release test pipeline:
- Upload working directory to Azure
- Upload metrics/results json file to Azure
- Download files (metrics/results json file) from Azure
- Helper function to parse ABFSS URI into account, container, and path

This PR is broken down from https://github.com/ray-project/ray/pull/57252 which also includes a sample hello world test on Azure to test e2e. Proof that it works: https://buildkite.com/ray-project/release/builds/62278

### CODE REVIEW

File: release/ray_release/command_runner/_anyscale_job_wrapper.py [hunk 1]
```diff
@@ -88,6 +88,17 @@ def run_storage_cp(source: str, target: str):
             source,
             target,
         ]
+    elif storage_service == "abfss":
+        subprocess.run(["azcopy", "login", "--identity"], check=True)
+        account, container, path = _parse_abfss_uri(target)
+        # Azcopy doesn't support ABFSS URI, so we need to convert it to a blob URI
+        new_target = f"https://{account}.dfs.core.windows.net/{container}/{path}"
+        cp_cmd_args = [
```
Comment (khluu) [inline release/ray_release/command_runner/_anyscale_job_wrapper.py:93]: This script is running by itself (without Bazel) so there's no Azure dependencies, and `azcopy` is available so it's used here

File: release/ray_release/command_runner/_anyscale_job_wrapper.py [hunk 2]
```diff
@@ -420,3 +431,27 @@ def main(
 
     args = parser.parse_args()
     sys.exit(main(**args.__dict__))
+
+
+def _parse_abfss_uri(uri: str) -> Tuple[str, str, str]:
```
Comment (khluu) [inline release/ray_release/command_runner/_anyscale_job_wrapper.py]: this function is also available in `ray_release/util.py` but this script doesn't run with Bazel so it doesn't recognize the util module.

Comment (aslonnie) [reply release/ray_release/command_runner/_anyscale_job_wrapper.py]: could you add this information in comment?

File: release/ray_release/command_runner/_anyscale_job_wrapper.py [hunk 3]
```diff
@@ -88,6 +88,17 @@ def run_storage_cp(source: str, target: str):
             source,
             target,
         ]
+    elif storage_service == "abfss":
+        subprocess.run(["azcopy", "login", "--identity"], check=True)
+        account, container, path = _parse_abfss_uri(target)
```
Comment (aslonnie) [inline release/ray_release/command_runner/_anyscale_job_wrapper.py]: taked offline, let's send something that is easier to use/parse, so that we do not need to have a tricky URL parser in this thin wrapper code.

Comment (khluu) [reply release/ray_release/command_runner/_anyscale_job_wrapper.py]: done

File: release/ray_release/buildkite/step.py [hunk 4]
```diff
@@ -154,6 +154,11 @@ def get_step(
     default_project_id = env_dict.get("RELEASE_DEFAULT_PROJECT")
     env_dict["ANYSCALE_PROJECT"] = get_test_project_id(test, default_project_id)
 
+    if test.is_azure():
+        env_dict["AZURE_CLIENT_ID"] = os.environ.get("AZURE_CLIENT_ID")
+        env_dict["AZURE_CLIENT_SECRET"] = os.environ.get("AZURE_CLIENT_SECRET")
+        env_dict["AZURE_TENANT_ID"] = os.environ.get("AZURE_TENANT_ID")
```
Comment (aslonnie) [inline release/ray_release/buildkite/step.py]: can we check if these are actually set here? and if not, throw an exception?

File: release/ray_release/command_runner/anyscale_job_runner.py [hunk 5]
```diff
@@ -76,6 +80,11 @@ def __init__(
             f"{cloud_storage_provider}://{self.file_manager.bucket}",
             self.path_in_bucket,
         )
+        if cloud_storage_provider == AZURE_CLOUD_STORAGE:
```
Comment (aslonnie) [inline release/ray_release/command_runner/anyscale_job_runner.py:80]: could you guard the default path in `else`, so that `join_cloud_storage_paths` is not executed twice for azure cloud storage?

and also add a comment for why azure's storage path require special handling?

File: release/ray_release/command_runner/anyscale_job_runner.py [hunk 6]
```diff
@@ -227,17 +236,40 @@ def run_command(
...
+        )
+        metrics_cloud_storage_uri = join_cloud_storage_paths(
+            self.upload_path, self._METRICS_OUTPUT_JSON
+        )
+        output_cloud_storage_uri = join_cloud_storage_paths(
+            self.upload_path, self.output_json
+        )
+        upload_cloud_storage_uri = self.upload_path
+        if self.upload_path.startswith(AZURE_CLOUD_STORAGE):
+            results_cloud_storage_uri = convert_abfss_uri_to_https(
```
Comment (aslonnie) [inline release/ray_release/command_runner/anyscale_job_runner.py:258]: could you add comment saying this is for `azcopy` to work?

File: release/ray_release/file_manager/job_file_manager.py [hunk 7]
```diff
@@ -75,7 +80,21 @@ def download_from_cloud(
             bucket = self.gs_client.bucket(self.bucket)
             blob = bucket.blob(key)
             self._run_with_retry(lambda: blob.download_to_filename(target))
+        if self.cloud_storage_provider == AZURE_CLOUD_STORAGE:
+            from azure.storage.blob import BlobServiceClient
+            from azure.identity import DefaultAzureCredential
```
Comment (aslonnie) [inline release/ray_release/file_manager/job_file_manager.py]: we do not need to do late import here, just import at the beginning of the file.

in general, try to avoid late import, it can lead to weird patterns of initialization sequence that is hard to understand / debug..

File: release/ray_release/file_manager/job_file_manager.py [hunk 8]
```diff
@@ -75,7 +80,21 @@ def download_from_cloud(
...
+            credential = DefaultAzureCredential(
+                exclude_managed_identity_credential=True
+            )
+            blob_service_client = BlobServiceClient(account_url, credential)
+            blob_client = blob_service_client.get_blob_client(
+                container=AZURE_STORAGE_CONTAINER, blob=key
+            )
+            with open(target, "wb") as f:
+                download_stream = blob_client.download_blob()
+                f.write(download_stream.readall())
```
[the remark this answers is not in the document]
Comment (aslonnie) [reply release/ray_release/file_manager/job_file_manager.py]: this is a good suggestion. please give it a try.

File: release/ray_release/test.py [hunk 9]
```diff
@@ -388,6 +388,12 @@ def is_kuberay(self) -> bool:
         """
         return self.get("env") == "kuberay"
 
+    def is_azure(self) -> bool:
```
Comment (aslonnie) [inline release/ray_release/test.py:391]: I feel that this is becoming too verbose now.. maybe should unit all those `is_some_provider()` calls into one `get_env()` that returns an enum or something.

File: release/ray_release/util.py [hunk 10]
```diff
@@ -239,6 +243,77 @@ def upload_working_dir(working_dir: str) -> str:
...
+        account, container, path = _parse_abfss_uri(azure_file_path)
+        account_url = f"https://{account}.blob.core.windows.net"
+        credential = DefaultAzureCredential(exclude_managed_identity_credential=True)
+        blob_service_client = BlobServiceClient(account_url, credential)
+        blob_client = blob_service_client.get_blob_client(
+            container=container, blob=path
+        )
+        with open(local_file_path, "rb") as f:
+            data = f.read()
+            blob_client.upload_blob(data=data)
```
[the remark this answers is not in the document]
Comment (aslonnie) [reply release/ray_release/util.py]: nice suggestion, maybe try this.

File: release/ray_release/util.py [hunk 11]
```diff
@@ -240,6 +244,86 @@ def upload_working_dir(working_dir: str) -> str:
...
+    except Exception as e:
+        logger.exception(f"Failed to upload file to Azure Blob Storage: {e}")
```
Comment (aslonnie) [inline release/ray_release/util.py:269]: why do we need this annotation?

if we really want this, can we just `try except` the `upload_blob` call, but leave other lines outside this `try` block?

File: release/ray_release/util.py [hunk 12]
```diff
@@ -240,6 +244,86 @@ def upload_working_dir(working_dir: str) -> str:
...
+    Args:
+        working_dir: Path to directory to upload.
+        azure_directory_uri: Path to directory in Azure blob storage.
+    Returns:
+        Azure blob storage path where archived directory was uploaded.
+    """
+    timestamp = str(int(time.time()))
+    archived_filename = f"ray_release_{timestamp}.zip"
+    output_path = os.path.abspath(archived_filename)
+    archived_file_path = shutil.make_archive(output_path[:-4], "zip", working_dir)
```
Comment (aslonnie) [inline release/ray_release/util.py]: these lines are duplicated with `upload_working_dir_to_gcs` ? could you extract them out into a helper function?

File: release/ray_release/util.py [hunk 13]
```diff
@@ -240,6 +244,86 @@ def upload_working_dir(working_dir: str) -> str:
...
+    azure_file_path = f"{azure_directory_uri}/{archived_filename}"
+    upload_file_to_azure(
+        local_file_path=archived_file_path, azure_file_path=azure_file_path
+    )
+    return azure_file_path
+
+
+def _parse_abfss_uri(uri: str) -> tuple:
```
Comment (aslonnie) [inline release/ray_release/util.py]: please add unit tests for these parser functions. try to hit the corner cases.

File: release/requirements_buildkite.in [hunk 14]
```diff
@@ -2,12 +2,15 @@
 # Copy anyscale pin to requirements.txt and util.py
 aioboto3
 anyscale >= 0.26.14
+azure-identity
+azure-storage-blob
```
Comment (aslonnie) [inline release/requirements_buildkite.in:6]: sync master I guess?

**Comment by aslonnie** [comment]
also, ci unit tests seem to be failing.

File: release/ray_release/util.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (khluu) [reply release/ray_release/util.py]: ah I did this intentionally so I could unit test it

File: release/ray_release/buildkite/step.py (hunk 4 above)
Comment (khluu) [reply release/ray_release/buildkite/step.py]: done

File: release/ray_release/command_runner/anyscale_job_runner.py (hunk 5 above)
Comment (khluu) [reply release/ray_release/command_runner/anyscale_job_runner.py:80]: done

File: release/ray_release/command_runner/anyscale_job_runner.py (hunk 6 above)
Comment (khluu) [reply release/ray_release/command_runner/anyscale_job_runner.py:258]: done

File: release/ray_release/util.py (hunk 11 above)
Comment (khluu) [reply release/ray_release/util.py:269]: done

File: release/ray_release/util.py (hunk 13 above)
Comment (khluu) [reply release/ray_release/util.py]: done

File: release/ray_release/buildkite/step.py (hunk 4 above)
Comment (aslonnie) [reply release/ray_release/buildkite/step.py]: it needs to check first? do something in the loop that:

````
for k in ["....."]:
    value = os.environ.get(k)
    if not value:
         raise ...
    env_dict[k] = value
````

File: release/ray_release/util.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (aslonnie) [reply release/ray_release/util.py]: why the other way is not unit testable? could you write a fake client with the same signature?

File: release/ray_release/util.py [hunk 15]
```diff
@@ -239,6 +243,77 @@ def upload_working_dir(working_dir: str) -> str:
...
```
[the remark this answers is not in the document]
Comment (aslonnie) [reply release/ray_release/util.py]: yes please.

File: release/ray_release/file_manager/job_file_manager.py (hunk 7 above)
Comment (aslonnie) [reply release/ray_release/file_manager/job_file_manager.py]: still here?

**Review by aslonnie — COMMENTED** [review]
(reviewed

File: release/ray_release/util.py [hunk 16]
```diff
@@ -240,6 +241,91 @@ def upload_working_dir(working_dir: str) -> str:
...
+    """Upload a file to Azure Blob Storage.
+
+    Args:
+        local_file_path: Path to local file to upload.
+        azure_file_path: Path to file in Azure blob storage.
+    """
+
+    # Import here because not every jobs that use ray_release/util need azure
+    from azure.storage.blob import BlobServiceClient
+    from azure.identity import DefaultAzureCredential
```
Comment (aslonnie) [inline release/ray_release/util.py]: please move these to the header of the file.

**Review by aslonnie — APPROVED** [review]
